### PR TITLE
output_area: Add expand/collapse icon in overlay prompt.

### PR DESCRIPTION
**Pull request by brijsiyag**
## References

#14341 

## Code changes
Added a `initialize` signal in `outputarea` and connected in `cells` package, to update the icon initially.

## User-facing changes

Yet to discuss about what should be icon's size or it should hide if output is a single line.

Screenshot 2023-06-05 at 7 29 12 AMScreenshot 2023-06-05 at 7 29 28 AMScreenshot 2023-06-05 at 7 30 02 AMScreenshot 2023-06-05 at 7 30 23 AMScreenshot 2023-06-05 at 7 31 07 AMScreenshot 2023-06-05 at 7 31 22 AM

### CODE REVIEW

**Comment by GabrielaVives** [comment]
Hi @brijsiyag , it would be best if the icon does not change size. I would suggest using a similar size to the other icons on the interface, probably the size that is used for the 1-line output. And for single line outputs I think we should hide the icon, as there is no scrolling available.

File: packages/outputarea/src/widget.ts
Comment (krassowski) [inline packages/outputarea/src/widget.ts:26]: Commented out code needs to be removed

File: packages/cells/src/widget.ts
Comment (krassowski) [inline packages/cells/src/widget.ts:1302]:
```suggestion
   * Update the Prompt Overlay Icon
```

Should the method name include word "prompt" too?

**Review by krassowski — COMMENTED** [review]
Thank you for working on this. Some suggestions on the code attached.

**Comment by krassowski** [comment]
Ther arrow icon is already used on collapsible headers which correspond to summary/details. This PR only shows icons on hover for switching scrolling mode which is distinct. Users would only see them when hovering over the area.

**Comment by tonyfast** [comment]
> Ther arrow icon is already used on collapsible headers which correspond to summary/details. This PR only shows icons on hover for switching scrolling mode which is distinct. Users would only see them when hovering over the area.

it doesn't seem like there needs to be a different icon if it is doing the job of opening and closing an element. the icon meaning in this implementation  is not apparent when you see it. so when hovering something confusing is exposed. 

my main question is there a better iconography for this behavior?

File: packages/ui-components/style/icons/output/collapse.svg
Comment (brijsiyag) [reply packages/ui-components/style/icons/output/collapse.svg:7]: Okay, I put in <g> tag, and removed stroke property. Also fixed precision. I think we don't have to handle the theme separately , as the color(From figma) of svg is suitable for both themes.
Thanks for review, I didn't gave attention to svg.

File: packages/outputarea/src/widget.ts
Comment (brijsiyag) [reply packages/outputarea/src/widget.ts:26]: Sure

**Comment by krassowski** [comment]
But it is not opening/closing an element here. It toggles the scroll mode.

File: packages/cells/src/widget.ts
Comment (brijsiyag) [reply packages/cells/src/widget.ts:1302]: Upated to `updatePromptOverlayIcon`.

**Comment by brijsiyag** [comment]
> Hi @brijsiyag , it would be best if the icon does not change size. I would suggest using a similar size to the other icons on the interface, probably the size that is used for the 1-line output. And for single line outputs I think we should hide the icon, as there is no scrolling available.

Should we set it to 16px ? I think 16px will be too small if we see figma design. I'm setting height 24px (Sidebar icon's size) for now, tell me if have to alter that.

File: packages/outputarea/style/base.css
Comment (brijsiyag) [reply packages/outputarea/style/base.css]: Used class name without child combinator, as svg is not a direct child.

**Comment by brijsiyag** [comment]
@krassowski , I have made the changes, don't know why one test is failing. PTAL !!

**Comment by GabrielaVives** [comment]
> > Hi @brijsiyag , it would be best if the icon does not change size. I would suggest using a similar size to the other icons on the interface, probably the size that is used for the 1-line output. And for single line outputs I think we should hide the icon, as there is no scrolling available.
> 
> Should we set it to 16px ? I think 16px will be too small if we see figma design. I'm setting height 24px (Sidebar icon's size) for now, tell me if have to alter that.

Let's try that and see how it feels. 24px is also a good icon height for desktop application so it might be ok!

**Comment by krassowski** [comment]
The release test is failing because the integrity script adds fallback CSS variables to `packages/ui-components/style/deprecated.css` for the icons newly added in this PR. Running `jlpm run integrity` locall should update that file (I am not sure why it is not picked up by integrity jobs); if it does not work you can just manually apply the diff:

<details>
 
```diff --git a/packages/ui-components/style/deprecated.css b/packages/ui-components/style/deprecated.css
index 9363378..d7b6d12 100644
--- a/packages/ui-components/style/deprecated.css
+++ b/packages/ui-components/style/deprecated.css
@@ -35,6 +35,7 @@
   --jp-icon-code-check: url('icons/lsp/code-check.svg');
   --jp-icon-code: url('icons/toolbar/code.svg');
   --jp-icon-collapse-all: url('icons/toolbar/collapse-all.svg');
+  --jp-icon-collapse: url('icons/output/collapse.svg');
   --jp-icon-console: url('icons/filetype/console.svg');
   --jp-icon-copy: url('icons/toolbar/copy.svg');
   --jp-icon-copyright: url('icons/licenses/copyright.svg');
@@ -46,6 +47,7 @@
   --jp-icon-ellipses: url('icons/toolbar/ellipses.svg');
   --jp-icon-error: url('icons/error.svg');
   --jp-icon-expand-all: url('icons/toolbar/expand-all.svg');
+  --jp-icon-expand: url('icons/output/expand.svg');
   --jp-icon-extension: url('icons/sidebar/extension.svg');
   --jp-icon-fast-forward: url('icons/toolbar/fast-forward.svg');
   --jp-icon-file-upload: url('icons/toolbar/file-upload.svg');
@@ -207,6 +209,10 @@
   background-image: var(--jp-icon-collapse-all);
 }
 
+.jp-CollapseIcon {
+  background-image: var(--jp-icon-collapse);
+}
+
 .jp-ConsoleIcon {
   background-image: var(--jp-icon-console);
 }
@@ -251,6 +257,10 @@
   background-image: var(--jp-icon-expand-all);
 }
 
+.jp-ExpandIcon {
+  background-image: var(--jp-icon-expand);
+}
+
 .jp-ExtensionIcon {
   background-image: var(--jp-icon-extension);
 }
```

</details>

**Comment by brijsiyag** [comment]
Is there any doc on what action/test does what ?

**Comment by andrii-i** [comment]
@brijsiyag Please see [Visual Regression and UI Tests](https://jupyterlab.readthedocs.io/en/latest/developer/contributing.html#id9) part of the jupyterlab docs, part "2. An intended update to user interface". When UI changes are done, you can just post the command to update galata snapshots here in the thread.

**Comment by brijsiyag** [comment]
> @brijsiyag Thank you for working on this. Please see [Visual Regression and UI Tests](https://jupyterlab.readthedocs.io/en/latest/developer/contributing.html#visual-regression-and-ui-tests) part of the jupyterlab docs, part "2. An intended update to user interface". When UI changes are done, you can post the command to update galata snapshots here in the thread.

This will help a lot in future, Thanks for sharing.

**Comment by brijsiyag** [comment]
please update galata snapshots

**Comment by brijsiyag** [comment]
bot please update galata snapshots

**Comment by andrii-i** [comment]
@brijsiyag bot picked up the request but workflow takes a while to complete. You can check Actions tab of the jupyterlab repo to see workflows that are currently running:
Screenshot 2023-06-09 at 1 32 57 PM

**Comment by brijsiyag** [comment]
Ooh, I thought it would post a message that it started the action. Sorry I kicked it three times.
Thank you for informing !!

**Comment by andrii-i** [comment]
@brijsiyag No worries, it took me some time to figure out as well.

**Comment by krassowski** [comment]
We definitely should make it leave an emoji reactions like the dependabot does. I opened https://github.com/jupyterlab/maintainer-tools/issues/201 to track this.

Here however, since this PR only adds on hover icon, is the snapshot update necessary? Or maybe a test was just flaky?

**Comment by andrii-i** [comment]
@krassowski that's a good point. Re-running the failed test.

**Comment by andrii-i** [comment]
Below is a GIF that illustrates how icon is shown on-hover only. Note that icon is hidden when output is 1-line high.

This PR adds a visual element that people working with large outputs would see often. Taking in account discussion with @tonyfast above, I think it would be nice to demo this change during (next) JupyterLab team meeting before merging it. 

@brijsiyag JupyterLab team meetings are held every Wednesday on Zoom. You can check the [Jupyter Community Meetings calendar](https://docs.jupyter.org/en/latest/community/content-community.html#jupyter-community-meetings) for the exact time in your timezone and Zoom link. If you can't make it, me or someone else in the thread who can bring it up during the call as well. 

icon_hover

**Comment by andrii-i** [comment]
Looking at the failed run (https://github.com/jupyterlab/jupyterlab/actions/runs/5225799071/jobs/9435721681) and report inside it, test/jupyterlab/debugger.test.ts:121 seems is flakey. It fails on the first run and passes on the retry:

 
Screenshot 2023-06-09 at 3 34 21 PM

**Comment by JasonWeill** [comment]
Thank you so much for contributing this pull request! I had added the custom zoom-in and zoom-out cursors because we didn't have a custom icon, but now that you're adding one, it might be good to remove the custom cursor CSS rules.

**Comment by andrii-i** [comment]
Also taking in account we have an icon at the center of the div now, do we need to change the hover color of the entire div? When we had zoom-in and zoom-out cursors only, this was necessary to indicate interaction area.

**Comment by andrii-i** [comment]
Also taking in account we have an icon at the center of the div now, do we need to change the hover color of the entire div? When we had zoom-in and zoom-out cursors only, this was necessary to indicate interaction area. 

Please note that this and other additional improvements can also be introduced later in separate PRs by other contributors, not necessarily in this one.

**Comment by andrii-i** [comment]
Also taking in account we have an icon at the center of the div now, do we need to change the hover color of the entire div? When we had zoom-in and zoom-out cursors only, this was necessary to indicate interaction area.

@brijsiyag Please note that this and other additional improvements can be additionally discussed and introduced later, not necessarily in this PR.

**Comment by krassowski** [comment]
I think hover colour is still useful as it indicates interaction boundaries locally. Alternatively we could change border outline colour on hover. In any case I would defer to a separate PR as suggested above by Andrii.

**Comment by tonyfast** [comment]
in cell mode we can use <kbd>Left/Right</kbd> to expand and collapse markdown cells. there are no keyboard <kbd>Left/Right</kbd> behaviors defined on code cells. would it be consistent to use <kbd>Left/Right</kbd> as a keyboard toggle for the expanded/collapsed state on the code cell?

**Comment by brijsiyag** [comment]
> Thank you so much for contributing this pull request! I had added the custom zoom-in and zoom-out cursors because we didn't have a custom icon, but now that you're adding one, it might be good to remove the custom cursor CSS rules.

Okay, I will add this and what @andrii-i suggested to my to-do cleanup list. Or I can open issues as well if you suggest.

I will try to be at JupyterLab team meeting or will notify here if unable to do so.

Thank you for your positive feedback.

**Comment by brijsiyag** [comment]
> in cell mode we can use Left/Right to expand and collapse markdown cells. there are no keyboard Left/Right behaviors defined on code cells. would it be consistent to use Left/Right as a keyboard toggle for the expanded/collapsed state on the code cell?

This is interesting, Have you submitted an issue already ?

**Comment by andrii-i** [comment]
@GabrielaVives the GIF in my post above (https://github.com/jupyterlab/jupyterlab/pull/14640#issuecomment-1585174327) is current state of this PR and you are correct that there is a change of color of the entire div when you hover over it.

I questioned if we need to change color of the entire div after addition of the icon. I think that changing border outline color only on hover vs color of the whole div. That would make UI look less cluttered while maintaining the same level of readability (this can be done in a separate PR). Would love to hear your opinion on this.

**Comment by tonyfast** [comment]
it might be easier to understand what the icons are doing if the text title save "Expand Output" / "Collapse Output" instead of "Toggle output".

**Comment by brijsiyag** [comment]
Updated the title on hover to what @tonyfast suggested,

Follow-ups after this PR merged will be- 
* Remove zoom-in/out cursor.
* Don't change color of entire div on hover, (Yet to decide but will implement this to have a look.)
* Use keyboard shortcut `Left/Right` to expand or collapse the output, (Needs further discussion.)S

File: packages/outputarea/style/base.css
Comment (fcollonval) [inline packages/outputarea/style/base.css:86]: Could you drop the cursor in this PR? It is just a matter of removing the cursor style rules line 92 and line 96 below.

File: packages/ui-components/style/icons/output/collapse.svg
Comment (fcollonval) [inline packages/ui-components/style/icons/output/collapse.svg:1]: Our icon must be squared. Could you please update the SVG and also should have the same default size `width="16"`?

Would you be also able to minimify the icon by merging the transformation with the path? If you don't know how to, I use the `jock.svg` extension for VS Code.

File: packages/ui-components/style/icons/output/expand.svg
Comment (fcollonval) [inline packages/ui-components/style/icons/output/expand.svg:1]: Same comment as the other icon.

**Review by fcollonval — CHANGES_REQUESTED** [review]
Thanks @brijsiyag

I have minor comments and this should be good to go.

File: packages/ui-components/style/icons/output/collapse.svg
Comment (brijsiyag) [reply packages/ui-components/style/icons/output/collapse.svg:1]: > > > Hi @brijsiyag , it would be best if the icon does not change size. I would suggest using a similar size to the other icons on the interface, probably the size that is used for the 1-line output. And for single line outputs I think we should hide the icon, as there is no scrolling available.
> > 
> > 
> > Should we set it to 16px ? I think 16px will be too small if we see figma design. I'm setting height 24px (Sidebar icon's size) for now, tell me if have to alter that.
> 
> Let's try that and see how it feels. 24px is also a good icon height for desktop application so it might be ok!

This is the comment from @GabrielaVives. Let me post a look of both here for comparison, 

Screenshot 2023-06-16 at 12 18 42 AM
Screenshot 2023-06-16 at 12 19 22 AM

@fcollonval PTAL and let me know which size we should use.

Comment (brijsiyag) [reply packages/ui-components/style/icons/output/collapse.svg:1]: and thanks for the extension suggestion, I will do update suggested changes.

Comment (krassowski) [reply packages/ui-components/style/icons/output/collapse.svg:1]: After seeing them side-by-side I would vote for 16px. Another point is that we can also resize icons using `LabIcon` attributes, but regardless of the size in UI, our SVG icon files are conventionally all 16px wide on 24x24 grid for consistency.

Comment (GabrielaVives) [reply packages/ui-components/style/icons/output/collapse.svg]: Same as @krassowski , I think the 16px looks consistent with the visual information surrounding this button. The larger version feels out of place.

Comment (brijsiyag) [reply packages/ui-components/style/icons/output/collapse.svg]: Yeah agreed, I've updated the PR with width="16".

File: packages/cells/src/widget.ts
Comment (fcollonval) [inline packages/cells/src/widget.ts:1313]: Last question @brijsiyag, should this be `16` following the icon size change?

> Thanks a lot for the patience

```suggestion
    const ICON_HEIGHT = 16 + 4 + 4; // 4px for padding
```

Comment (brijsiyag) [reply packages/cells/src/widget.ts:1313]: Ooh sorry, yes this need to be 16, thanks for informing. I will update it.

**Comment by brijsiyag** [comment]
Failed tests don't seem related to this PR.

bot please update galata snapshots

**Review by fcollonval — APPROVED** [review]
Thanks a lot for the patience @brijsiyag

**Pull request merged**
